keps/sig-cloud-provider/2395-removing-in-tree-cloud-providers/README.md

```diff
@@ -195,6 +236,191 @@ import (
+functionality is accessed.
+
+[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+- Initial e2e tests completed and enabled
+
+#### Beta
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Additional tests are in Testgrid and linked in KEP
+
+#### GA
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+TBD
+
+### Upgrade / Downgrade Strategy
+
+<!--
+If applicable, how will the component be upgraded and downgraded? Make sure
+this is in the test plan.
+
+Consider the following in developing an upgrade/downgrade strategy for this
+enhancement:
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to maintain previous behavior?
+- What changes (in invocations, configurations, API use, etc.) is an existing
+  cluster required to make on upgrade, in order to make use of the enhancement?
+-->
+
+TBD
+
+### Version Skew Strategy
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and nodes?
+- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
+- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+TBD
+
 ## Production Readiness Review Questionnaire
 
 ### Feature Enablement and Rollback
```

Comment:
On `Can the feature be disabled once it has been enabled (i.e. can we roll back
the enablement)?`, what efforts have been taken to make sure the new code doesn't make changes that would cause the old code to misbehave?  Do we have any tests that try to exercise this fallback?

Reply:
The `Rollout, Upgrade and Rollback Planning` section need to be filled out.

Same for `Monitoring Requirements`, `Dependencies`, `Troubleshooting`.

Reply:
The `Scalability` section isn't correct.  For example this answer:

> No, it will actually remove calls to the cloud provider in all core components.

misses the point of the question completely.  We aren't asking in relation to who is making the calls, just that the calls are being made.  So what effort has been taken to make sure that the new code doesn't create a bunch of extra load that the old code didn't?

Reply:
>  what efforts have been taken to make sure the new code doesn't make changes that would cause the old code to misbehave?

if i understand your question correctly, the e2e tests that are currently in place will ensure that the old code is not active.

>  Do we have any tests that try to exercise this fallback?

i do not think we have any testing which exercises the rollback. i'll have to talk with the sig more about this, but this testing will be extremely difficult as it requires the core controllers to be reconfigured and then restarted. i don't think we had scoped this type of testing, i will double check though.

> The Rollout, Upgrade and Rollback Planning section need to be filled out.
> Same for Monitoring Requirements, Dependencies, Troubleshooting.

i'm still working on these, i left a note in my latest comment but i'd like to get more help from the sig answering these.

> misses the point of the question completely. We aren't asking in relation to who is making the calls, just that the calls are being made. So what effort has been taken to make sure that the new code doesn't create a bunch of extra load that the old code didn't?

i did not write this answer, but the answer is factual from the core component perspective. this change will remove many calls to cloud providers that currently exist in the kubernetes code. but, it also changes that focus to the external cloud provider implementations.

in terms of the effort that has been taken to ensure we aren't creating extra load, that effort is being moved to the external CCMs which will be maintained by their individual communities. the common cloud-provider framework that the sig maintains does make improvements to the way the cloud controllers batch calls to the providers, but each CCM project is free to choose their implementation. 

one thing is true though, there is no way for the new code to create more load on the cloud provider APIs than the old code as the new code is a removal of all the old code. i suppose the confirmation of this is that tests which require cloud controllers simply will not work if there are no external CCMs engaged, there is no cloud to call as none of that code exists without the external CCM running.

Reply:
> That is not what I mean. Say a field in a resource was always set to 5 with the legacy code, and always set to 6 by the new code. On a downgrade, the resource would still exist, but now the old code would see a 6, which it may not understand or act correctly against.

i'm having some trouble following this example. i'd like to talk through it.

in the case of this KEP, if a user upgrades to the new behavior, they will be disabling some loops that run in kubelet/kas/kcm and running them as external pods provided by their cloud infrastructure provider (or whichever team is maintaining, but not core kubernetes). if they  upgrade then they are changing flags in kubelet/kas/kcm and adding pods to the cluster, if they downgrade then they are re-enabling those flags and terminating the pods they were running. following from your example, let's say that the in-tree (old code) modifies a Node object with some value (the theoretical `5`), then the user upgrade to external (new code) and the value get modified (to a `6`), finally they downgrade back to the original code and now there is an error.  am i getting closer to your intention?

afaik, the former in-tree cloud controllers which became the external CCMs are not modifying Node or Service objects in a way that would create this incompatibility, but it is certainly possible. our reliance here has been on the behavioral tests in the e2e suite to confirm that Kubernetes continues to operate as expected. with that said though, due to the nature of testing specific cloud providers on their own infrastructure this testing has always been limited.

> I don't see how this can be true. A new controller could easily have a hot loop bug that the old controller did not. So by upgrading, the new controller would create way more load against the cloud provider than before, even if everything works correctly in CI.

perhaps i'm just misunderstanding, part of my disconnect here is that this KEP describes a process where users will need to source an external CCM for their clusters depending on the cloud provider they are running on. it is certainly possible that a user could run a controller which hot loops and creates extra load but i don't think it's in scope for this KEP to cover all the possible controllers that could be created. when this KEP was in alpha state, all the code from the in-tree controllers was moved into out-of-tree repositories that are maintained by provider specific teams. it is possible that in the interim those projects have added more to their controller loops, but as i understand our test infrastructure we have not been testing previous controllers on all the providers due to resource constraints (e.g. donated hardware/cloud time). given the testing that is currently done for in-tree cloud controllers, i'm not sure how we would answer the call to test all the possible controllers or even just test the controllers that were previously in-tree. 

fwiw, we (sig cloud provider) are working towards improving the testing situation for all cloud providers, but as i understand it currently we only deeply test the cloud controllers on gce, with a subset of those tests running on aws, and a subset of those running on azure. in the future we will have a testing suite that providers can run from their external CCM repositories to more deeply test any provider who wishes to participate.

back to your original question:

> So what effort has been taken to make sure that the new code doesn't create a bunch of extra load that the old code didn't?

we have just enabled the external CCMs by default in the kubernetes CI, i'm not sure we have specific tests there which show the differences in load but i will ask around to see if there is something we could use as a reference. if not, we will need to develop those tests.

Reply:
@enj i updated the answers to the api call questions to better reflect what is happening and might happen in the future. does this help?